controllers/kuadrant_status.go

```diff
@@ -0,0 +1,85 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	"github.com/kuadrant/kuadrant-operator/pkg/common"
+)
+
+const (
+	ReadyConditionType string = "Ready"
+)
+
+func (r *KuadrantReconciler) reconcileStatus(ctx context.Context, kObj *kuadrantv1beta1.Kuadrant, specErr error) (ctrl.Result, error) {
+	logger := logr.FromContext(ctx)
+	newStatus := r.calculateStatus(kObj, specErr)
+
+	equalStatus := kObj.Status.Equals(newStatus, logger)
+	logger.V(1).Info("Status", "status is different", !equalStatus)
+	logger.V(1).Info("Status", "generation is different", kObj.Generation != kObj.Status.ObservedGeneration)
+	if equalStatus && kObj.Generation == kObj.Status.ObservedGeneration {
+		// Steady state
+		logger.V(1).Info("Status was not updated")
+		return reconcile.Result{}, nil
+	}
+
+	// Save the generation number we acted on, otherwise we might wrongfully indicate
+	// that we've seen a spec update when we retry.
+	// TODO: This can clobber an update if we allow multiple agents to write to the
+	// same status.
+	newStatus.ObservedGeneration = kObj.Generation
+
+	logger.V(1).Info("Updating Status", "sequence no:", fmt.Sprintf("sequence No: %v->%v", kObj.Status.ObservedGeneration, newStatus.ObservedGeneration))
+
+	kObj.Status = *newStatus
+	updateErr := r.Status().Update(ctx, kObj)
+	if updateErr != nil {
+		// Ignore conflicts, resource might just be outdated.
+		if errors.IsConflict(updateErr) {
+			logger.Info("Failed to update status: resource might just be outdated")
+			return reconcile.Result{Requeue: true}, nil
+		}
+
+		return reconcile.Result{}, fmt.Errorf("failed to update status: %w", updateErr)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *KuadrantReconciler) calculateStatus(kObj *kuadrantv1beta1.Kuadrant, specErr error) *kuadrantv1beta1.KuadrantStatus {
+	newStatus := &kuadrantv1beta1.KuadrantStatus{
+		// Copy initial conditions. Otherwise, status will always be updated
+		Conditions: common.CopyConditions(kObj.Status.Conditions),
+	}
+
+	availableCond := r.readyCondition(specErr)
+
+	meta.SetStatusCondition(&newStatus.Conditions, *availableCond)
+
+	return newStatus
+}
+
+func (r *KuadrantReconciler) readyCondition(specErr error) *metav1.Condition {
+	cond := &metav1.Condition{
```

Comment:
`metav1.Condition.LastTransitionTime` is annotated as `+kubebuilder:validation:Required`. Do we need to add it here?

https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition